File: apps/mac-ear/Sources/VegaEar/SilenceDetector.swift
import Foundation

// Streaming VAD with auto-calibrated noise floor. The first
// `calibrationMs` of each session are used to estimate the ambient RMS
// (75th percentile so a single keypress click doesn't poison the floor).
// After that, speech is detected when RMS rises `speechMargin` above the
// floor, and silence is declared when it drops within `silenceMargin` of
// the floor. Endpoint fires after `endSilenceMs` of sustained silence
// following observed speech.

final class SilenceDetector {
    struct Config {
        var endSilenceMs: Int = 3_000
        var graceMs: Int = 500
        var calibrationMs: Int = 600
        var speechMargin: Double = 300
        var silenceMargin: Double = 100
        var fallbackNoiseFloor: Double = 100
    }

    private let config: Config
    private var sawSpeech = false
    private var silenceStartedAt: Date?
    private let startedAt = Date()
    private var noiseFloorRms: Double = 0
    private var calibrationSamples: [Double] = []
    private var lastLoggedState: String = ""
    // When suppressEndpoint is true, the detector keeps computing RMS for
    // logging but never returns .endpoint. Used by long-note mode.
    private var suppressEndpoint = false

    func setEndpointSuppressed(_ suppressed: Bool) {
        if suppressEndpoint != suppressed {
            NSLog("[VegaEar] VAD endpoint suppression \(suppressed ? "ON" : "OFF")")
        }
        suppressEndpoint = suppressed
    }

    init(config: Config = Config()) {
        self.config = config
    }

    enum Decision {
        case waiting
        case ongoing
        case endpoint
    }

    func feed(pcm: Data) -> Decision {
        let nowMs = Date().timeIntervalSince(startedAt) * 1000
        let rms = Self.computeRms(pcm)

        if nowMs < Double(config.calibrationMs) {
            calibrationSamples.append(rms)
            return .waiting
        }
        if noiseFloorRms == 0 {
            if calibrationSamples.isEmpty {
                noiseFloorRms = config.fallbackNoiseFloor
            } else {
                let sorted = calibrationSamples.sorted()
                let idx = max(0, min(sorted.count - 1, Int(Double(sorted.count) * 0.75)))
                noiseFloorRms = max(config.fallbackNoiseFloor, sorted[idx])
            }
            NSLog(String(
                format: "[VegaEar] VAD calibrated: noiseFloor=%.0f speechThr=%.0f silenceThr=%.0f",
                noiseFloorRms,
                noiseFloorRms + config.speechMargin,
                noiseFloorRms + config.silenceMargin
            ))
        }

        if nowMs < Double(config.graceMs) {
            return .waiting
        }

        let speechThresh = noiseFloorRms + config.speechMargin
        let silenceThresh = noiseFloorRms + config.silenceMargin

        if rms >= speechThresh {
            if !sawSpeech {
                NSLog(String(format: "[VegaEar] VAD speech detected, RMS=%.0f threshold=%.0f", rms, speechThresh))
            }
            sawSpeech = true
            silenceStartedAt = nil
            logState("speech")
            return .ongoing
        }
        if !sawSpeech {
            logState("pre-speech")
            return .waiting
        }
        if rms <= silenceThresh {
            if silenceStartedAt == nil {
                silenceStartedAt = Date()
                NSLog(String(format: "[VegaEar] VAD silence started, RMS=%.0f threshold=%.0f", rms, silenceThresh))
            }
            if let started = silenceStartedAt,
               Date().timeIntervalSince(started) * 1000 >= Double(config.endSilenceMs) {
                if suppressEndpoint {
                    // Don't fire endpoint in long-note mode; let safety cap own termination.
                    logState("silence-suppressed")
                } else {
                    NSLog("[VegaEar] VAD endpoint: sustained silence \(config.endSilenceMs) ms")
                    return .endpoint
                }
            }
            logState("silence")
        } else {
            if silenceStartedAt != nil {
                NSLog(String(format: "[VegaEar] VAD silence broken, RMS=%.0f", rms))
            }
            silenceStartedAt = nil
            logState("speech-tail")
        }
        return .ongoing
    }

    private func logState(_ state: String) {
        // Rate-limit by only logging state transitions (not every frame).
        guard state != lastLoggedState else { return }
        lastLoggedState = state
    }

    private static func computeRms(_ pcm: Data) -> Double {
        let count = pcm.count / MemoryLayout<Int16>.size
        guard count > 0 else { return 0 }
        var sumSquares: Double = 0
        pcm.withUnsafeBytes { raw in
            let buf = raw.bindMemory(to: Int16.self)
            for s in buf {
                let f = Double(s)
                sumSquares += f * f
            }
        }
        return sqrt(sumSquares / Double(count))
    }
}

File: apps/core/src/session/silence-detector.ts
// Server-side VAD that mirrors the Ear's adaptive silence detector.
// First `calibrationMs` of each session estimate the noise floor (75th
// percentile of observed RMS); afterwards speech is detected when RMS
// rises ≥ `speechMargin` above the floor, silence when it sits within
// `silenceMargin` of the floor. Endpoint fires after `endSilenceMs` of
// sustained silence following observed speech.

export interface SilenceDetectorOptions {
  endSilenceMs: number;
  graceMs: number;
  calibrationMs: number;
  speechMargin: number;
  silenceMargin: number;
  fallbackNoiseFloor: number;
}

export const DEFAULT_SILENCE_OPTS: SilenceDetectorOptions = {
  endSilenceMs: 5_000,
  graceMs: 500,
  calibrationMs: 600,
  speechMargin: 300,
  silenceMargin: 100,
  fallbackNoiseFloor: 100,
};

export type SilenceDecision = "waiting" | "ongoing" | "endpoint";

export class SilenceDetector {
  private readonly startedAt = Date.now();
  private noiseFloorRms = 0;
  private calibrationSamples: number[] = [];
  private sawSpeech = false;
  private silenceStartedAt: number | null = null;
  private calibrationLogged = false;
  private speechLogged = false;

  constructor(
    private readonly opts: SilenceDetectorOptions = DEFAULT_SILENCE_OPTS,
    private readonly onLog?: (msg: string, meta?: Record<string, unknown>) => void,
  ) {}

  feed(pcm: Uint8Array): SilenceDecision {
    const nowMs = Date.now() - this.startedAt;
    const rms = SilenceDetector.computeRms(pcm);

    if (nowMs < this.opts.calibrationMs) {
      this.calibrationSamples.push(rms);
      return "waiting";
    }

    if (this.noiseFloorRms === 0) {
      if (this.calibrationSamples.length === 0) {
        this.noiseFloorRms = this.opts.fallbackNoiseFloor;
      } else {
        const sorted = [...this.calibrationSamples].sort((a, b) => a - b);
        const idx = Math.min(sorted.length - 1, Math.floor(sorted.length * 0.75));
        this.noiseFloorRms = Math.max(this.opts.fallbackNoiseFloor, sorted[idx]);
      }
      if (!this.calibrationLogged) {
        this.calibrationLogged = true;
        this.onLog?.(
          `VAD calibrated`,
          {
            noiseFloor: Math.round(this.noiseFloorRms),
            speechThr: Math.round(this.noiseFloorRms + this.opts.speechMargin),
            silenceThr: Math.round(this.noiseFloorRms + this.opts.silenceMargin),
          },
        );
      }
    }

    if (nowMs < this.opts.graceMs) {
      return "waiting";
    }

    const speechThr = this.noiseFloorRms + this.opts.speechMargin;
    const silenceThr = this.noiseFloorRms + this.opts.silenceMargin;

    if (rms >= speechThr) {
      if (!this.sawSpeech && !this.speechLogged) {
        this.speechLogged = true;
        this.onLog?.(`VAD speech detected`, { rms: Math.round(rms), threshold: Math.round(speechThr) });
      }
      this.sawSpeech = true;
      this.silenceStartedAt = null;
      return "ongoing";
    }

    if (!this.sawSpeech) return "waiting";

    if (rms <= silenceThr) {
      if (this.silenceStartedAt === null) {
        this.silenceStartedAt = Date.now();
        this.onLog?.(`VAD silence started`, { rms: Math.round(rms), threshold: Math.round(silenceThr) });
      }
      if (Date.now() - this.silenceStartedAt >= this.opts.endSilenceMs) {
        this.onLog?.(`VAD endpoint reached`, { silenceMs: this.opts.endSilenceMs });
        return "endpoint";
      }
    } else {
      if (this.silenceStartedAt !== null) {
        this.onLog?.(`VAD silence broken`, { rms: Math.round(rms) });
      }
      this.silenceStartedAt = null;
    }
    return "ongoing";
  }

  get currentNoiseFloor(): number {
    return this.noiseFloorRms;
  }

  private static computeRms(pcm: Uint8Array): number {
    const count = Math.floor(pcm.byteLength / 2);
    if (count === 0) return 0;
    const view = new DataView(pcm.buffer, pcm.byteOffset, pcm.byteLength);
    let sumSquares = 0;
    for (let i = 0; i < count; i++) {
      const s = view.getInt16(i * 2, true);
      sumSquares += s * s;
    }
    return Math.sqrt(sumSquares / count);
  }
}
